### PR TITLE
1700 push zero

### LINKED ISSUE #1700: Add PUSH0 instruction to EVM
Add support for PUSH0 in Shanghai fork.

Solidity compiler version 0.8.20 uses lots of PUSH0 instructions by default

### PR DESCRIPTION
**Pull request by kladkogex**
This has been reviewed already

Path timestamp: `pushZeroPatchTimestamp `
See the tests in tests/historicstate/hardhat/scripts/push0_test.ts

### CODE REVIEW

**Review by dimalit — DISMISSED** [review]

File: test/historicstate/hardhat/scripts/push0_test.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        throw message;
+    }
+}
+
+async function getAndPrintTrace(hash: string): Promise<String> {
+//    const trace = await ethers.provider.send('debug_traceTransaction', [hash, {"tracer":"prestateTracer",
+//        "tracerConfig": {"diffMode":true}}]);
+
+//    const trace = await ethers.provider.send('debug_traceTransaction', [hash, {"tracer": "callTracer",
+//        "tracerConfig": {"withLog":true}}]);
```
Comment (DmytroNazarenko) [inline test/historicstate/hardhat/scripts/push0_test.ts]: Please, remove commented lines of code if not needed or uncomment them

File: test/historicstate/hardhat/scripts/push0_test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    const deployBn = await ethers.provider.getBlockNumber();
+
+    const hash = testContract.deployTransaction.hash;
+    console.log(`Gas limit ${testContract.deployTransaction.gasLimit}`);
+    console.log(`Contract deployed to ${testContract.address} at block ${deployBn} tx hash ${hash}`);
+
+
+    // await waitUntilNextBlock()
+
+    //await getAndPrintTrace(hash)
```
Comment (DmytroNazarenko) [inline test/historicstate/hardhat/scripts/push0_test.ts]: These too

File: libevm/LegacyVM.cpp [hunk 3]
```diff
@@ -1355,6 +1356,21 @@ void LegacyVM::interpretCases() {
         }
         CONTINUE
 
+        // EIP-3855. Code PUSH0 is similar to PUSH1 but pushes 0
+        // we need to increment program counter only by one since
+        // the value is not read from program code as in PUSH1
+        CASE( PUSH0 ) {
+            if ( !PushZeroPatch::isEnabled() ) {
```
Comment (DmytroNazarenko) [inline libevm/LegacyVM.cpp:1363]: Will it work in the same way on non-updated nodes and updated nodes before patch will be enabled?

File: libconsensus
Comment (DmytroNazarenko) [inline libconsensus:1]: Do we need to update consensus for this fix?

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]
Minor changes requested. Let's create a custom build and check it on the devnet before merge

**Review by sergiy-skalelabs — APPROVED** [review]

**Review by DmytroNazarenko — APPROVED** [review]
Verified on devnet

**Pull request merged**